### ISSUE TITLE
Support > Is it possible to use QueryInterfaceRequest builder with dynamic table/view name?

### ISSUE THREAD

**Issue by Kirow** [groue/GRDB.swift#990]
### What did you do?

let's say I have following data model:

```swift
struct MyRecord: FetchableRecord {
    let title: String
    let details: String
    
    enum Columns: String, CodingKey, ColumnExpression {
        case title, details
    }
    
    init(row: Row) {
        title = row[Columns.title]
        details = row[Columns.details]
    }
}
```

And I'd like to use `QueryInterfaceRequest` to build queries and load records. But in my case I don't have table name at compile time and single model should be used across multiple tables. So it seems I can not adopt `TableRecord` and as result have no access to `QueryInterfaceRequest` builder.

for example lets say I'm loading table definition from some external source:
```swift
struct Definition {
    let name: String
    let keys: [String]
}

let myDefinitions = [
    Definition(name: "table1", keys: ["key1", "key2"]),
    Definition(name: "asdasd", keys: ["whatever", "whocares"])
]
```

### What did you expect?

I'm looking for some way to create `QueryInterfaceRequest<MyRecord>` using provided `Definition` and existing `MyRecord: FetchableRecord` 

for example:

```swift
// technically it is working for me if I change access level inside GRDB
let requests: [QueryInterfaceRequest<MyRecord>] = myDefinitions.map({ def in
    let rel = SQLRelation.all(fromTable: def.name, selection: { _ in [
        Column(def.keys[0]).forKey(MyRecord.Columns.title),
        Column(def.keys[1]).forKey(MyRecord.Columns.details),
    ] })
    
    return QueryInterfaceRequest<MyRecord>(relation: rel)
})
```

### Current workaround

If it is not possible to use `QueryInterfaceRequest` in this case, then I will be using custom wrapper with raw SQL.

```swift
struct MyQueryWrapper<T: FetchableRecord> {
    let sql: (_ limit: Int, _ offset: Int) -> String
    
    func fetch(_ db: Database, limit: Int, offset: Int) throws -> [T] {
        return try T.fetchAll(db, sql: sql(limit, offset))
    }
}

let wrappers = myDefinitions.map({ def in
    return MyQueryWrapper<MyRecord> { limit, offset in
        """
            SELECT
                \(def.keys[0]) AS \(MyRecord.Columns.title.name),
                \(def.keys[1]) AS \(MyRecord.Columns.details.name)
            FROM \(def.name)
            LIMIT \(limit) OFFSET \(offset)
        """
    }
})
```

### Environment

**GRDB flavor(s):** GRDB
**GRDB version:** 5.8.0
**Installation method:** CocoaPods, manual
**Xcode version:** 12.5
**Swift version:** 5.4
**Platform(s) running GRDB:** iOS, macOS
**macOS version running Xcode:** 11.2.3

### Demo Project

<details>
<summary>Playground</summary>

```swift
import GRDB

// default database config, nothing interesting
var configuration = Configuration()
configuration.prepareDatabase { db in
    db.trace { print("SQL> \($0)") }
}
let dbQueue = DatabaseQueue(configuration: configuration)


// model to be decoded from different tables
struct MyRecord: FetchableRecord {
    let title: String
    let details: String
    
    enum Columns: String, CodingKey, ColumnExpression {
        case title, details
    }
    
    init(row: Row) {
        title = row[Columns.title]
        details = row[Columns.details]
    }
}

struct Definition {
    let name: String
    let keys: [String]
}

let myDefinitions = [
    Definition(name: "table1", keys: ["key1", "key2"]),
    Definition(name: "asdasd", keys: ["whatever", "whocares"])
]


// create tables, add records
try dbQueue.write { db in
    
    for def in myDefinitions {
        // create random tables
        try db.create(table: def.name) { t in
            t.autoIncrementedPrimaryKey("id")
            for key in def.keys {
                t.column(key, .text).notNull()
            }
        }
        
        // some records
        for index in 1...10 {
            let rawSQL = "INSERT INTO \(def.name) (\(def.keys.joined(separator: ", "))) VALUES ('\(index)', '\(UUID().uuidString)')"
            try db.execute(sql: rawSQL)
        }
    }
}


// current workaround
struct MyQueryWrapper<T: FetchableRecord> {
    let sql: (_ limit: Int, _ offset: Int) -> String
    
    func fetch(_ db: Database, limit: Int, offset: Int) throws -> [T] {
        return try T.fetchAll(db, sql: sql(limit, offset))
    }
}

// wrappers from definitions
let wrappers = myDefinitions.map({ def in
    return MyQueryWrapper<MyRecord> { limit, offset in
        """
            SELECT
                \(def.keys[0]) AS \(MyRecord.Columns.title.name),
                \(def.keys[1]) AS \(MyRecord.Columns.details.name)
            FROM \(def.name)
            LIMIT \(limit) OFFSET \(offset)
        """
    }
})


// use test
try dbQueue.read { db in
    let result1 = try wrappers[0].fetch(db, limit: 2, offset: 2)
    print(result1)
    let result2 = try wrappers[1].fetch(db, limit: 2, offset: 2)
    print(result2)
}

/* Limited Access
 
let requests: [QueryInterfaceRequest<MyRecord>] = myDefinitions.map({ def in
    let rel = SQLRelation.all(fromTable: def.name, selection: { _ in [
        Column(def.keys[0]).forKey(MyRecord.Columns.title),
        Column(def.keys[1]).forKey(MyRecord.Columns.details),
    ] })
    
    return QueryInterfaceRequest<MyRecord>(relation: rel)
})


// use test
try dbQueue.read { db in
    let result1 = try requests[0].limit(2, offset: 2).fetchAll(db)
    print(result1)
    let result2 = try requests[1].limit(2, offset: 2).fetchAll(db)
    print(result2)
}

*/
```

</details>

**Comment by groue** [comment]
Hello @Kirow,

You are not the first one to ask for a way to build QueryInterfaceRequest with a dynamic table name: #916, #833, maybe other issues.

I used to be reluctant, but I'm no longer. There clearly is a demand for this. And there already exist a branch with preliminary support for this feature.

Consider that you have just bumped the topic, and that the feature will ship in the next GRDB release.

**Comment by groue** [comment]
QueryInterfaceRequest builder with dynamic table/view name is addressed by the pull request #993.

**Comment by Kirow** [comment]
Thanks for update! It does provide functionality that I was missing for my specific case.

```swift
let requests: [QueryInterfaceRequest<MyRecord>] = myDefinitions.map({ def in
    return Table<MyRecord>(def.name).select([
        Column(def.keys[0]).forKey(MyRecord.Columns.title),
        Column(def.keys[1]).forKey(MyRecord.Columns.details),
    ])
})
```

**Comment by groue** [comment]
Thanks for the feedback, @Kirow 👍

**Comment by groue** [comment]
`Table` has shipped in v5.9.0 - [documentation](https://github.com/groue/GRDB.swift/blob/v5.9.0/README.md#requests)
